benchmarks/REPS_CREPS_gradients/creps.py

```diff
@@ -0,0 +1,344 @@
+    """Solve dual function for C-REPS.
+
+    Parameters
+    ----------
+    S : array, shape (n_samples_per_update, n_context_features)
+        Features for the context-dependend reward baseline
+
+    R : array, shape (n_samples_per_update,)
+        Corresponding obtained rewards
+
+    epsilon : float
+        Maximum Kullback-Leibler divergence of two successive policy
+        distributions.
+
+    min_eta : float
+        Minimum eta, 0 would result in numerical problems
+
+    Returns
+    -------
+    d : array, shape (n_samples_per_update,)
+        Weights for training samples
+
+    eta : float
+        Temperature
+
+    nu : array, shape (n_context_features,)
+        Coefficients of linear reward baseline function
+    """
+    if S.shape[0] != R.shape[0]:
+        raise ValueError("Number of contexts (%d) must equal number of "
+                         "returns (%d)." % (S.shape[0], R.shape[0]))
+
+    n_samples_per_update = len(R)
+
+    # Definition of the dual function
+    def g(x):  # Objective function
+        eta = x[0]
+        nu = x[1:]
+        return (eta * epsilon + nu.T.dot(S.mean(axis=0)) +
+                eta * logsumexp((R - nu.dot(S.T)) / eta,
+                    b=1.0 / n_samples_per_update))
+
+    def g_grad(x):  # Objective function
+        eta = x[0]
+        nu = x[1:]
+
+        s_mean = S.mean(axis=0)
+        R_over_eta = (R - nu.dot(S.T)) / eta
+        log_sum_exp = logsumexp(R_over_eta, b=1.0 / n_samples_per_update)
+        Z = np.exp(R_over_eta - R_over_eta.max())
+        Z_sum = Z.sum()
+
+        f = eta * (epsilon + log_sum_exp) + nu.T.dot(s_mean)
+        d_eta = epsilon + log_sum_exp - Z.dot(R_over_eta) / Z_sum
+        d_nu = s_mean - Z.dot(S) / Z_sum
+
+        return f, np.append(d_eta, d_nu)
+
+    # Lower bound for Lagrange parameters eta and nu
+    bounds = np.vstack(([[min_eta, None]], np.tile(None, (S.shape[1], 2))))
+    # Start point for optimization
+    x0 = [1] + [1] * S.shape[1]
+
+    # Perform the actual optimization of the dual function
+    #r = NLP(g, x0, lb=lb).solve('ralg', iprint=-10)
+    if approx_grad:
+        r = fmin_l_bfgs_b(g, x0, approx_grad=True, bounds=bounds)
+    else:
+        r = fmin_l_bfgs_b(g_grad, x0, bounds=bounds)
+
+    # Fetch optimal lagrangian parameter eta. Corresponds to a temperature
+    # of a softmax distribution
+    eta = r[0][0]
+    # Fetch optimal vale of vector nu which determines the context
+    # dependent baseline
+    nu = r[0][1:]
+
+    # Determine weights of individual samples based on the their return,
+    # the optimal baseline nu.dot(\phi(s)) and the "temperature" eta
+    log_d = (R - nu.dot(S.T)) / eta
+    # Numerically stable softmax version of the weights. Note that
+    # this does neither changes the solution of the weighted least
+    # squares nor the estimation of the covariance.
+    d = np.exp(log_d - log_d.max())
+    d /= d.sum()
+
+    return d, eta, nu
+
+
+class CREPSOptimizer(ContextualOptimizer):
```

Comment:
The problem with copying the whole file is that if there are any changes in (c)reps in the future, you have to copy them to the benchmark. You should make a subclass of CREPSOptimizer that overrides `set_evaluation_feedback`. The same should be done for the REPSOptimizer.